### ISSUE TITLE
[rfc] [pipelining] shape inference + cached buffer allocation

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#136811]
TLDR: we plan to do shape inference, so users don't have to supply input/output shapes/dtypes to each pipeline stage they create.  (this is purely a convenience feature, but a helpful one).   After doing shape inference, we can delete shape validation code.

We also want to minimize the number of recv buffers we allocate, which is easy to do after adding shape inference code and deleting shape-validation code.

Implementation ideas:

buffer allocation
- delete the code that creates n_microbatch buffers up front
- this means also updating (or deleting) validate_fwd_shapes in PipelineStageBase
   - rather than updating this, prefer to do shape-inference first and then delete this

shape-inference

```
stage:

   def perform_shape_inference(self):
      # knows self stage id
      # has 'stage to rank' map, which determines whether 'next stage' is on the same rank or not

      # 1: recv input shapes
      # if prev stage is not on the same rank, issue a recv op
      # else, expect shapes are passed into this function call?  or what?

      # 2: call forward
      # use 'no_grad' context

      # 3: propagate output shapes
      # if next stage is colocated, return output shapes
      # else, perform send

   
schedule:

   def step(args, kwargs...)

      # perform shape inference:
      # this can be a base-schedule API and it should work for loop/V schedules becuase its only assumption is that
      # the stages should propagate the shape in the order of the stage indices
      maybe_input_shapes = input_shapes
      for s in self.stages:
         # perform_shape_infernece will return None if it sent shapes to the next stage via nccl, otherwise
         # return the shapes to pass to the next stage that's on the same rank (special case for V schedule)
         maybe_input_shapes = s.perform_shape_inference(maybe_input_shapes)

      # run the actual schedule...
```

One debate was over whether we should do shape inference 'separately' or 'embedded inside the schedule'.  

'Separately' seems simpler to implement, but runs an extra forward pass that is observable, and we'd ideally not do that.  We mitigate by running under 'no_grad()' context but there is still a risk that someone registered forward hooks that we'd end up calling, or something like that.

builds on work from @H-Huang at https://github.com/pytorch/pytorch/pull/130856

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

**Comment by H-Huang** [comment]
Looks good to me. A few clarifying questions on the pseudocode, but feel free to code and it will probably be answered through that:

Is `perform_shape_inference` also going to set the buffers on the stage? Or does that happen in `step()`

`perform_shape_inference` probably is going to have dependency on the `dist.send/recv_object_list` right? Or did you have another idea for which collective to use.

`for s in self.stages:` I realized we already make an assumption in our existing code that stages are passed as a list and the stage_indices are in correct, increasing order. So we should validate that, or reorder them so that they always are?

**Comment by kwen2501** [comment]
LGTM.
We may be also able to use this mechanism to change pipeline buffer size dynamically, for example:
```
schedule.step(big_tensor)
schedule.reset_input_size()
schedule.step(small_tensor)
```

**Comment by kwen2501** [comment]
> for s in self.stages: I realized we already make an assumption in our existing code that stages are passed as a list and the stage_indices are in correct, increasing order. So we should validate that, or reorder them so that they always are?

I thought about that too. Otherwise there will be a hang.
We should force user to give them in an increasing order, I guess? Otherwise there will be a hang.
Should be something easy to check during Schedule init.

**Comment by kwen2501** [comment]
> there is still a risk that someone registered forward hooks that we'd end up calling

Yeah, I agree this is the thing we need to be careful about. In particular if the hook user adds would change something stateful.

**Comment by wconstab** [comment]
> Is perform_shape_inference also going to set the buffers on the stage? Or does that happen in step()

shape inference would only set the inputs/outputs meta info on the stage.  This would make it at parity with a stage that was initialized with user-provided inputs/outputs meta.  There would be no buffers allocated until later during step().  This coincides with step() also being smart about minimal allocations.  Instead of up-front allocation of `n_microbatches` buffers, we allocate the minimal buffers we need based on how many recv ops happen to overlap in the schedule.

> perform_shape_inference probably is going to have dependency on the dist.send/recv_object_list right? Or did you have another idea for which collective to use.

I think I would start by using send_object_list, but i would make the tensors 'meta tensors' before sending them so there is no 'pickling of tensor data' which is expensive.

> for s in self.stages: I realized we already make an assumption in our existing code that stages are passed as a list and the stage_indices are in correct, increasing order. So we should validate that, or reorder them so that they always are?

I think this is more than an assumption today- its an API requirement.  isn't it?  or else it is a strong assumption we should probably clarify.  Even the schedule itself assumes this though.

**Comment by H-Huang** [comment]
Resolved in https://github.com/pytorch/pytorch/pull/136912
